synapse/handlers/register.py

```diff
@@ -274,11 +274,18 @@ def check_user_id_is_valid(self, user_id):
                 )
 
     def generate_token(self, user_id):
-        # urlsafe variant uses _ and - so use . as the separator and replace
-        # all =s with .s so http clients don't quote =s when it is used as
-        # query params.
-        return (base64.urlsafe_b64encode(user_id).replace('=', '.') + '.' +
-                stringutils.random_string(18))
+        macaroon = pymacaroons.Macaroon(
+            location = self.hs.config.server_name,
+            identifier = "key",
+            key = self.hs.config.macaroon_secret_key)
+        macaroon.add_first_party_caveat("gen = 1")
+        macaroon.add_first_party_caveat("user_id = %s" % user_id)
+        macaroon.add_first_party_caveat("type = access")
+        now = self.hs.get_clock().time()
+        expiry = now + 60 * 60
+        macaroon.add_first_party_caveat("time < %s" % expiry)
```

Comment:
- We generally always use milliseconds in our APIs. What are the benefits of using seconds?
- Generally I try and use `%d` for numbers instead of `%s` to make it clear what we expect and to avoid fun surprises if `expiry` happens to magically be something other than a number.
- `expiry` should be in a tuple, i.e. `"time < %s" % (expiry,)` (note the trailing `,`). Not putting it in a tuple *usually* works, but it can lead to surprised e.g. if printing a list the result changes depending on if you have 1 or more items in the list.